Reimplementation of drop/(uninstall)/remove

mih opened the pull request:
Sits on top of #6110 and will grow further. For now just a demo of the state for the curious.

Please help with naming! We need labels for

- `--what`
   - keys detetermined by paths in the worktree (default operation): `filecontent`?
   - to engage `git annex drop --auto`: `unwanted`?
   - to engage `git annex drop --all`: `allkeys`?
   - to uninstall entire datasets: `all`?

- `--reckless`
   - to disable minimum key redundancy checks: `availability`
   - to disable test if all branches are pushed: `availability` (yes, same as above -- same recklessness different subject only. one matters for file drop, the other for dataset uninstall), @adswa gave example, when the two should be selectable individually, though
   - to disable checks for unsaved modifications: `modification`
   - to disable announcement/push of `git annex dead here` outcome: `undead`

- Fixes #1142
- Fixes #1282
- Fixes #1823
- Fixes #1953
- Fixes #2229
- Fixes #2328
- Fixes #2655
- Fixes #2673
- Fixes #2967
- Fixes #3887
- Fixes #4097
- Fixes #4115
- Fixes #4784
- Fixes #5842
- Fixes #6123 

TODO:

- [x] drop in replacement for the previous `drop` implementation
- [x] remove old `drop` implementation
- [x] 95+% percent test coverage, using a new set of tests that are confirmed to work cross-platform
- [x] support passing `--jobs` to `git annex drop` #1953
- [x] support dropping all keys from all branches #2328 (criticial precondition for safely uninstalling datasets)
- [x] drop in replacement for `uninstall`
- [x] decide if with `drop --what all --reckless availability` we should even attempt to drop all keys. We could simply wipe out the repo, and be done (I feel like we discussed this before). already possible to bypass with `kill`
- [x] should the outcome of `git annex dead here` be pushed to the default remote (if any), or all remotes? What if one of them fails? What if it needs a merge first?
       No! If, and only if, the existence of a local annex was ever communicated elsewhere (most would not), an error message must inform users what to do to communicate its pending death. But not automatically inside `drop`.
- [x] check for potential undead annex records #3887
- [x] same for all branches: should they be pushed to the default remote (if any), or all remotes? What if one of them fails? What if it needs a merge first?
      Same no! An error message should inform about this, and instruct what to do (i.e. datalad push), but not do anything automatically inside `drop`
- [x] verify branch state availability before dropping an entire dataset
- [x] can drop toplevel datasets #2967
- [x] go through use case list at http://docs.datalad.org/en/stable/design/drop.html and confirm/deny support for each
- [x] document commands
- [x] decide on "unpushed" detection with adjusted branches. I think we should `localsync()` before testing this, and then only report the corresponding branches. The adjusted ones are not meant to be pushed anyways.
- [x] bring back `kill`
   ![image](https://user-images.githubusercontent.com/136479/139197738-dd1b9971-046d-47f4-8718-add84ce3d83c.png)
- [x] `uninstall` is a thin shim of a command around `drop`, old tests pass with minimal adjustments for increased safety
- [x] rewrite `remove` as a thin wrapper around `drop` and `save`, and remove all old helper code.
- [x] made the last usage of path annotation obsolete and dropped the entire code collection that was required to do this previously
- [x] rename `remove(what=...`) to `remove(drop=...)`
- [x] PR against -deprecated to move `AnnotatePaths` there (`publish` is built on it) https://github.com/datalad/datalad-deprecated/pull/46
- [x] `reckless=kill` should maybe be `kill-dataset`, it won't actually kill anything, when given a file path to drop, or actually affect file dropping too

adswa commented on datalad/distributed/tests/test_drop.py:
```diff
@@ -0,0 +1,432 @@
+    # some windows test setup is not very robust, explicitly
+    # include the default name "origin" in the test success
+    # conditions to make this more robust
+    assert_in(res[0]['message'][1], ['dl-test-remote', 'origin'])
```
I think the reason for the test failure is that your testing something like
```
>>> ['ab'] in ['ab', 'cd']
False
```
when it actually should be
```
>>> 'ab' in ['ab', 'cd']
True
```

adswa replied:
4c88098 hopefully fixed this

mih replied:
Argh, stupid me. Thx!

adswa commented:
I hopefully pushed a fix for the failing test

adswa commented:
Judging from the currently running tests, the test in question is fixed. But I just tried on windows and 

```
eq_(res[0]['message'][1], ['dl-test-remote'])
```

(without the origin) works just as well. maybe this works across platforms, too

mih commented:
So the reason underlying the failing `remove` test is that `status` cannot see the removal, hence `save` doesn't act on it.

After `remove` ran, the subdataset in question is gone, and the subproject removal 
```
C:\DLTMP\datalad_temp__eocjulu>git status
On branch adjusted/master(unlocked)
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   three (new commits)
        deleted:    two

C:\DLTMP\datalad_temp__eocjulu>git diff HEAD -- two
diff --git a/two b/two
deleted file mode 160000
index beed3a2..0000000
--- a/two
+++ /dev/null
@@ -1 +0,0 @@
-Subproject commit beed3a2f14cbc57f8efe67de9a375eb81c809933

C:\DLTMP\datalad_temp__eocjulu>datalad status
nothing to save, working tree clean
```

This needs to be investigated.

On appveyor we have `git version 2.26.2.windows.1`, on another windows box with `git version 2.33.0.windows.2` the exact same test passes.

adswa commented:
I can trigger the failure, but onlz bz downgrading Git

```
C:\Users\datalad\repos\datalad>git --version
git version 2.33.0.windows.2

C:\Users\datalad\repos\datalad>python -m nose -s -v datalad/distribution/tests/test_uninstall.py:test_clean_subds_removal
datalad.distribution.tests.test_uninstall.test_clean_subds_removal ... subdataset(ok): one (dataset)
subdataset(ok): two (dataset)
subdataset(ok): two (dataset)
subdataset(ok): three (dataset)
subdataset(ok): two (dataset)
subdataset(ok): three (dataset)
subdataset(ok): two (dataset)
subdataset(ok): three (dataset)
ok
Versions: annexremote=1.5.0 appdirs=1.4.4 boto=2.49.0 cmd:7z=19.00 cmd:annex=8.20210804-g1d3f59a9d cmd:bundled-git=UNKNOWN cmd:git=2.33.0.windows.2 cmd:system-git=2.33.0.windows.2 cmd:system-ssh=for humanize=3.2.0 iso8601=0.1.14 keyring=23.0.0 keyrings.alt=UNKNOWN msgpack=1.0.2 requests=2.25.1 wrapt=1.12.1
Obscure filename: str=b";&%b5{}'.datc" repr=";&%b5{}'.datc"
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='cp1252'
Environment: PATH='C:\\Program Files\\Docker\\Docker\\Resources\\bin;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\WINDOWS\\System32\\OpenSSH\\;C:\\ProgramData\\chocolatey\\bin;C:\\Program Files (x86)\\Gpg4win\\..\\GnuPG\\bin;C:\\Program Files\\PuTTY\\;C:\\Program Files\\Git\\cmd;C:\\Users\\datalad\\AppData\\Roaming\\local\\bin;C:\\Users\\datalad\\AppData\\Local\\Programs\\Python\\Python39\\Scripts\\;C:\\Users\\datalad\\AppData\\Local\\Programs\\Python\\Python39\\;C:\\Users\\datalad\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files (x86)\\GnuWin32\\bin;C:\\Program Files\\7-Zip;;C:\\Users\\datalad\\AppData\\Local\\Programs\\Microsoft VS Code\\bin' GIT_CONFIG_PARAMETERS="'init.defaultBranch=dl-test-branch' 'clone.defaultRemoteName=dl-test-remote'"

----------------------------------------------------------------------
Ran 1 test in 137.071s

OK

C:\Users\datalad\repos\datalad>git --version
git version 2.26.0.windows.1

C:\Users\datalad\repos\datalad>python -m nose -s -v datalad/distribution/tests/test_uninstall.py:test_clean_subds_removal
datalad.distribution.tests.test_uninstall.test_clean_subds_removal ... subdataset(ok): one (dataset)
subdataset(ok): two (dataset)
subdataset(ok): two (dataset)
subdataset(ok): three (dataset)
subdataset(ok): two (dataset)
subdataset(ok): three (dataset)
subdataset(ok): two (dataset)
FAIL
Versions: annexremote=1.5.0 appdirs=1.4.4 boto=2.49.0 cmd:7z=19.00 cmd:annex=8.20210804-g1d3f59a9d cmd:bundled-git=UNKNOWN cmd:git=2.26.0.windows.1 cmd:system-git=2.26.0.windows.1 cmd:system-ssh=for humanize=3.2.0 iso8601=0.1.14 keyring=23.0.0 keyrings.alt=UNKNOWN msgpack=1.0.2 requests=2.25.1 wrapt=1.12.1
Obscure filename: str=b";&%b5{}'.datc" repr=";&%b5{}'.datc"
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='cp1252'
Environment: PATH='C:\\Program Files\\Docker\\Docker\\Resources\\bin;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\WINDOWS\\System32\\OpenSSH\\;C:\\ProgramData\\chocolatey\\bin;C:\\Program Files (x86)\\Gpg4win\\..\\GnuPG\\bin;C:\\Program Files\\PuTTY\\;C:\\Program Files\\Git\\cmd;C:\\Users\\datalad\\AppData\\Roaming\\local\\bin;C:\\Users\\datalad\\AppData\\Local\\Programs\\Python\\Python39\\Scripts\\;C:\\Users\\datalad\\AppData\\Local\\Programs\\Python\\Python39\\;C:\\Users\\datalad\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files (x86)\\GnuWin32\\bin;C:\\Program Files\\7-Zip;;C:\\Users\\datalad\\AppData\\Local\\Programs\\Microsoft VS Code\\bin' GIT_CONFIG_PARAMETERS="'init.defaultBranch=master' 'clone.defaultRemoteName=origin'"

======================================================================
FAIL: datalad.distribution.tests.test_uninstall.test_clean_subds_removal
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\datalad\AppData\Local\Programs\Python\Python39\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "C:\Users\datalad\repos\datalad\datalad\tests\utils.py", line 737, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "C:\Users\datalad\repos\datalad\datalad\distribution\tests\test_uninstall.py", line 113, in test_clean_subds_removal
    assert_in_results(
  File "C:\Users\datalad\repos\datalad\datalad\tests\utils.py", line 1325, in assert_in_results
    _check_results_in(True, results, **kwargs)
  File "C:\Users\datalad\repos\datalad\datalad\tests\utils.py", line 1318, in _check_results_in
    raise AssertionError(msg.format(_format_res(kwargs),
AssertionError: Desired result
  {
   "action": "remove",
   "path": "C:\\Users\\datalad\\AppData\\Local\\Temp\\datalad_temp_xahpplwj\\two"
  }
not found among
  [
   {
    "action": "drop",
    "message": [
     "nothing to drop from %s",
     "C:\\Users\\datalad\\AppData\\Local\\Temp\\datalad_temp_xahpplwj\\two"
    ],
    "path": "C:\\Users\\datalad\\AppData\\Local\\Temp\\datalad_temp_xahpplwj\\two",
    "refds": "C:\\Users\\datalad\\AppData\\Local\\Temp\\datalad_temp_xahpplwj",
    "status": "notneeded",
    "type": "directory"
   },
   {
    "action": "save",
    "path": "C:\\Users\\datalad\\AppData\\Local\\Temp\\datalad_temp_xahpplwj",
    "refds": "C:\\Users\\datalad\\AppData\\Local\\Temp\\datalad_temp_xahpplwj",
    "status": "notneeded",
    "type": "dataset"
   }
  ]
-------------------- >> begin captured logging << --------------------
datalad.utils: Level 5: Importing datalad.utils
datalad.utils: DEBUG: Failed to query or got useless SC_ARG_MAX sysconf, will use hardcoded value: module 'os' has no attribute 'sysconf'
datalad.utils: DEBUG: Maximal length of cmdline string (adjusted for safety margin): 16383
datalad.utils: Level 5: Done importing datalad.utils
datalad.runner.gitrunner: Level 9: Will use git under '' (no adjustments to PATH if empty string)
datalad.runner.runner: DEBUG: Run ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin'] (cwd=None)
datalad.runner.protocol: Level 8: Process 2164 started
datalad.runner.nonasyncrunner: Level 5: ReaderThread(<_io.FileIO name=4 mode='rb' closefd=True>, <queue.Queue object at 0x000002A9EBF6ECD0>, ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin']) started
datalad.runner.nonasyncrunner: Level 5: ReaderThread(<_io.FileIO name=3 mode='rb' closefd=True>, <queue.Queue object at 0x000002A9EBF6ECD0>, ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin']) started
datalad.runner.protocol: Level 5: Read 1024 bytes from 2164[stdout]
datalad.runner.protocol: Level 5: Read 604 bytes from 2164[stdout]
datalad.runner.nonasyncrunner: Level 5: ReaderThread(<_io.FileIO name=4 mode='rb' closefd=True>, <queue.Queue object at 0x000002A9EBF6ECD0>, ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin']) exiting (stream end)
datalad.runner.nonasyncrunner: Level 5: ReaderThread(<_io.FileIO name=3 mode='rb' closefd=True>, <queue.Queue object at 0x000002A9EBF6ECD0>, ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin']) exiting (stream end)
datalad.runner.protocol: Level 8: Process 2164 exited with return code 0
datalad.runner.runner: DEBUG: Finished ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin'] with status 0
datalad.ui: Level 5: Starting importing ui
datalad.ui.dialog: Level 5: Starting importing ui.dialog
datalad.ui.dialog: Level 5: Done importing ui.dialog
datalad.ui: Level 5: Initiating UI switcher
datalad.ui: DEBUG: UI set to DialogUI(out=<StreamWrapper>)
datalad.ui: Level 5: Done importing ui
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 136.252s

FAILED (failures=1)
```

mih commented:
Yes, same here -- not on Windows, but also on my Debian laptop. With Git 2.26.2 `status` does not see the change. Investigating...

adswa commented:
with ``--what allkeys`` I don't see any path reporting - is this intentional? I makes it hard to guess which file only has one copy

```
❱ datalad drop --what allkeys
drop(ok): . (key)
drop(error): . (key) [unsafe
Could only verify the existence of 0 out of 1 necessary copy
(Use --reckless to override this check, or adjust numcopies.)]
drop(ok): . (key)
action summary:
  drop (error: 1, ok: 2)
```

Compared to 

```
❱ datalad drop .
drop(error): archive.tar.gz (file) [unsafe; Could only verify the existence of 0 out of 1 necessary copy; (Use --reckless to override this check, or adjust numcopies.)]
drop(ok): archive/3 (file)
drop(ok): archive/5 (file)
drop(impossible): . (directory) [could not drop some content in /tmp/abcd ['/tmp/abcd/archive.tar.gz']]
action summary:
  drop (error: 1, impossible: 1, ok: 2)
```

mih commented:
> with `--what allkeys` I don't see any path reporting - is this intentional? I makes it hard to guess which file only has one copy

True. The key drop results are associated with the path of the dataset. In full, they look like this

```
% dl -f json_pp drop --what allkeys
{
  "action": "drop",
  "key": "MD5E-s4--ba1f2511fc30423bdbb183fe33f3dd0f",
  "message": "unsafe\nCould only verify the existence of 0 out of 1 necessary copy\n(Use --reckless to override this check, or adjust numcopies.)",
  "path": "/tmp/top",
  "refds": "/tmp/top",
  "status": "error",
  "type": "key"
}
```

We could think about a dedicated result renderer that searches for key usage. However, I general it is not guaranteed that the is just one file associated with a key, or any file with a path in the current worktree, or even any file at all (unused key).

adswa commented:
> We could think about a dedicated result renderer that searches for key usage. However, I general it is not guaranteed that there is just one file associated with a key, or any file with a path in the current worktree, or even any file at all (unused key).

Would it be an option to just not report paths then? Just a count of successful operations, for example?

The linebreak after ``unsafe`` looks a bit funky as well:

```
drop(error): HCP1200/996782 (key) [unsafe
Could only verify the existence of 0 out of 1 necessary copy
(Note that these git remotes have annex-ignore set: origin)
(Use --reckless to override this check, or adjust numcopies.)]
```

adswa commented:
Sorry for all the edge cases I'm digging up, feel free to ignore if they get too obscure:

The automatic hint to use ``--reckless`` when file availability checks need to be overridden reads a bit weird when I've used the ``--reckless parameter`` for a different type of reckless behavior:

e.g., here is a dataset with  untracked content, and unique content, and the check that warns about the lack of copies for the unique key points to the parameter I'm already using. 

```
❱ datalad drop HCP1200/996782/ --reckless modification --what datasets              
[WARNING] Failed to load entrypoint crawler: DistributionNotFound(The 'datalad_deprecated' distribution was not found and is required by the application) 
drop(error): HCP1200/996782 (key) [unsafe
Could only verify the existence of 0 out of 1 necessary copy
(Note that these git remotes have annex-ignore set: origin)
(Use --reckless to override this check, or adjust numcopies.)]
```

Would it make sense to specify ``--reckless availability`` in this hint?

adswa commented on datalad/distributed/drop.py:
```diff
@@ -0,0 +1,791 @@
+
+        if what in ('all', 'datasets') and reckless == 'kill' and not recursive:
+            raise ValueError(
+                'A reckless kill is requested by no recursion flag is set. '
```
```suggestion
                'A reckless kill is requested but no recursion flag is set. '
```

mih commented:
> Would it be an option to just not report paths then? Just a count of successful operations, for example?

The `path` property is the only indication to know from which dataset keys are being dropped (this could be a recursive operation).

> The linebreak after `unsafe` looks a bit funky as well:

This is a message from git-annex. it has the newline already. Not sure if we want to mess with it further, but we could.

> Would it make sense to specify `--reckless availability` in this hint?

It would, yes.

adswa commented:
Just a note to myself: I see that the new drop became even more careful. Previously, a newly created, empty dataset could be just removed, now I get a warning that the ``master`` branch isn't pushed anywhere:

```
(handbook2) adina@muninn in /tmp/abcd on git:master
❱ datalad drop --what datasets acb 
uninstall(error): acb (dataset) [to-be-dropped dataset has the following revisions that are not available at any known sibling. Use `datalad push --to ...` to push these before dropping the local dataset, or ignore via `--reckless availability`: ['master']]
(handbook2) adina@muninn in /tmp/abcd on git:master
❱ datalad remove -d acb                                                                                                           
uninstall(error): . (dataset) [to-be-dropped dataset has the following revisions that are not available at any known sibling. Use `datalad push --to ...` to push these before dropping the local dataset, or ignore via `--reckless availability`: ['master']]

```

mih commented:
> Just a note to myself: I see that the new drop became even more careful. Previously, a newly created, empty dataset could be just removed, now I get a warning that the `master` branch isn't pushed anywhere:

Yes, and on top of that: If you ever pushed a dataset elsewhere, it will complain about undead annexes too, and refuse dropping.

adswa commented on datalad/distributed/drop.py:
```diff
@@ -0,0 +1,791 @@
+            With 'undead', detection whether a to-be-removed local annex is
+            still known to exists in the network of dataset-clones is disabled.
+            This could cause zombie-records of invalid file availability.
+            With 'kill', all safety-checks are disabled.""",
```
When I use ``kill``, I get a value error for not having used ``--recursive`` as well. Maybe its useful to document this here, too?

Also, I learned that kill only operates on datasets, and wouldn't kill filecontent:

```
❱ datalad save
add(ok): file (file)                                                                                                                  
save(ok): . (dataset)                                                                                                                 
action summary:                                                                                                                       
  add (ok: 1)
  save (ok: 1)
(handbook2) adina@muninn in /tmp/abcd on git:master
❱ datalad drop --what filecontent --reckless kill file 
(handbook2) adina@muninn in /tmp/abcd on git:master
❱ ls
acb  file
```
Maybe this could be stated in the docstring, too? Or it could yield an "impossible" result when the path to be dropped is a file?

mih replied:
I am afraid the only clean approach would be to change the logic to also have `kill` impact dropping file content.

mih replied:
Done

mih commented on datalad/distributed/drop.py:
```diff
@@ -0,0 +1,791 @@
+                type='dataset',
+                status='error',
+                message=(
+                    "to-be-dropped dataset has the following revisions that "
+                    "are not available at any known sibling. Use "
+                    "`datalad push --to ...` to push "
+                    "these before dropping the local dataset, "
+                    "or ignore via `--reckless availability`: %s",
```
Makes sense! Thx!

adswa commented on datalad/distributed/drop.py:
```diff
@@ -0,0 +1,792 @@
+                 '(will fail with subdatasets present)',
+         'code_py': "drop(what='all')",
+         'code_cmd': "datalad drop --what all"},
+        {'text': 'Kill a dataset recklessly with any existing subdatasets too'
+                 '(this will be fast, but will disable any and all safety '
+                 'checks)',
```
```suggestion
        {'text': 'Kill a dataset recklessly with any existing subdatasets too '
                 '(this will be fast, but will disable any and all safety '
                 'checks)',
```

There is a missing space:

```
Kill a dataset recklessly with any existing subdatasets too(this will
be fast, but will disable any and all safety checks)::
```

adswa commented on datalad/local/remove.py:
```diff
@@ -0,0 +1,229 @@
+    for p in ('reckless', 'if_dirty', 'check',):
+        _params_[p] = Drop._params_[p]
+
+    _examples_ = [
+        dict(text="Permanently remove a subdataset from a dataset",
+             code_py="remove(dataset='path/to/dataset', path='path/to/subds')",
+             code_cmd="datalad remove -d <path/to/dataset> <path/to/subds>"),
+    ]
+
```
maybe emphasize the inherent recursiveness of operations in the examples? And maybe also mention removal of a superdataset as an example (it is my most used usecase for remove)

```suggestion
    _examples_ = [
        dict(text="Permanently remove a subdataset (and all further subdatasets contained in it) from a dataset",
             code_py="remove(dataset='path/to/dataset', path='path/to/subds')",
             code_cmd="datalad remove -d <path/to/dataset> <path/to/subds>"),
             dict(text="Permanently remove a superdataset (with all subdatasets) from the filesystem",
             code_py="remove(dataset='path/to/dataset')",
             code_cmd="datalad remove -d <path/to/dataset>"),
    ]

```

mih replied:
Makes sense. Thx!

adswa commented on datalad/local/remove.py:
```diff
@@ -0,0 +1,229 @@
+    where a minimal storage footprint for local datasets installations is
+    desirable.
+
+    Removing a dataset component is always a recursive operation. Removing a
+    directory, removes all content underneath the directory too. If
+    subdatasets are located under a to-be-removed path, they will be
+    uninstalled entirely, and all their content dropped. If any subdataset
+    can not be uninstalled safely, the remove operation will fail and halt.
+    """
```
Maybe this is the right place to mention that remove won't operate on ``.``?

mih replied:
Not sure. My view was that removing CWD would be extremely unexpected behavior -- hence confirming that it won't happen is not needed. Same rational as for UNIX `rm`. It is also not in the docs that it won't remove CWD, but still wouldn't.

adswa replied:
ok

adswa commented on datalad/local/remove.py:
```diff
@@ -0,0 +1,229 @@
+            lgr.debug('Observed drop failure, will not attempt remove')
+            return
+
+        for dpath, paths in paths_by_ds.items():
```
Not yet sure where this problem is rooted, but ``remove`` fails to save a subdataset deletion if it is a hierarchy if datasets:

```
# remove a single level of subdatasets is fine:
❱ datalad get -n HCP1200/996782   
[INFO   ] Fetching 'http://store.datalad.org/89d/4899e-1fe6-11ea-9e62-002590496000/config' 
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782 (dataset) [Installed subdataset in order to get /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782]
❱ datalad remove HCP1200/996782                                                                                                   
uninstall(ok): HCP1200/996782 (dataset)
remove(ok): HCP1200/996782 (file)                                                                                                     
save(ok): . (dataset)                                                                                                                 
action summary:                                                                                                                       
  remove (ok: 1)
  save (ok: 1)
  uninstall (ok: 1)
```


```
# removal of many levels of subdatasets leaves the superdataset dirty:
❱ datalad get -n -r HCP1200/996782

[INFO   ] Installing Dataset(/home/adina/repos/human-connectome-project-openaccess/HCP1200/996782) to get /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782 recursively 
[INFO   ] scanning for annexed files (this may take some time) 
[INFO   ] Fetching 'http://store.datalad.org/937/b2e76-1fe6-11ea-8fe3-002590496000/config' 
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782/MNINonLinear (dataset)                              
[INFO   ] Fetching 'http://store.datalad.org/94f/c881c-1fe6-11ea-8fe3-002590496000/config'                                            
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782/T1w (dataset)                                       
[INFO   ] Fetching 'http://store.datalad.org/96a/2d630-1fe6-11ea-8fe3-002590496000/config'                                            
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782/unprocessed (dataset)                               
action summary:                                                                                                                       
  get (notneeded: 1)
  install (ok: 3)
datalad get -n -r HCP1200/996782  5.83s user 3.39s system 32% cpu 27.943 total
(handbook2) adina@muninn in ~/repos/human-connectome-project-openaccess on git:master
❱ datalad remove -d HCP1200/996782
uninstall(ok): MNINonLinear (dataset)
uninstall(ok): T1w (dataset)
uninstall(ok): unprocessed (dataset)
uninstall(ok): . (dataset)
action summary:
  uninstall (ok: 4)
(handbook2) adina@muninn in ~/repos/human-connectome-project-openaccess on git:master!
❱ git st
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    HCP1200/996782
```

mih replied:
> `datalad remove -d HCP1200/996782`

You are not removing *from* the superdataset, hence it stays unsaved. Without `-d`:

```sh
(datalad-dev) mih@meiner /tmp/hcp (git)-[master] % datalad remove HCP1200/996782
uninstall(ok): HCP1200/996782/MNINonLinear (dataset)
uninstall(ok): HCP1200/996782/T1w (dataset)
uninstall(ok): HCP1200/996782/unprocessed (dataset)
uninstall(ok): HCP1200/996782 (dataset)
remove(ok): HCP1200/996782 (file)                                                                                           
save(ok): . (dataset)                                                                                                       
action summary:                                                                                                             
  remove (ok: 1)
  save (ok: 1)
  uninstall (ok: 4)
(datalad-dev) mih@meiner /tmp/hcp (git)-[master] % git st
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
```

adswa replied:
ehhh right :facepalm:

mih commented:
Good FS:

```
% python -m nose -s -v datalad.local.tests.test_remove datalad.distributed.tests.test_drop datalad.distribution.tests.test_uninstall  --with-coverage --cover-erase --cover-package datalad.local.remove --cover-package datalad.distributed.drop
Name                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------
datalad/distributed/drop.py     209      8    96%   189-190, 324-331, 403, 562, 607, 716
datalad/local/remove.py          56      0   100%
-----------------------------------------------------------
TOTAL                           265      8    97%
----------------------------------------------------------------------
Ran 30 tests in 109.508s

OK
```

Crippled FS (hence by-passing the often inappropriate `if-on-windows` conditionals:

```
======================================================================
FAIL: datalad.distribution.tests.test_uninstall.test_remove_file_handle_only
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mih/env/datalad-dev/lib/python3.9/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/mih/hacking/datalad/git/datalad/tests/utils.py", line 558, in _wrap_with_tree
    return t(*(arg + (d,)), **kw)
  File "/home/mih/hacking/datalad/git/datalad/distribution/tests/test_uninstall.py", line 292, in test_remove_file_handle_only
    eq_(rpath_one, (ds.pathobj / 'two').resolve())
AssertionError: PosixPath('/media/mih/Samsung_T5/tmp/datalad_temp_tree_test_remove_file_handle_onlyonew6nnh/one') != PosixPath('/media/mih/Samsung_T5/tmp/datalad_temp_tree_test_remove_file_handle_onlyonew6nnh/two')

======================================================================
FAIL: datalad.distribution.tests.test_uninstall.test_uninstall_multiple_paths
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mih/env/datalad-dev/lib/python3.9/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/mih/hacking/datalad/git/datalad/tests/utils.py", line 558, in _wrap_with_tree
    return t(*(arg + (d,)), **kw)
  File "/home/mih/hacking/datalad/git/datalad/distribution/tests/test_uninstall.py", line 248, in test_uninstall_multiple_paths
    ok_(all([f.endswith('keep') for f in files_left if exists(f) and not isdir(f)]))
AssertionError: None

======================================================================
FAIL: datalad.distribution.tests.test_uninstall.test_uninstall_recursive
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mih/env/datalad-dev/lib/python3.9/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/mih/hacking/datalad/git/datalad/tests/utils.py", line 558, in _wrap_with_tree
    return t(*(arg + (d,)), **kw)
  File "/home/mih/hacking/datalad/git/datalad/distribution/tests/test_uninstall.py", line 339, in test_uninstall_recursive
    ok_(not exists(lname))
AssertionError: None

Name                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------
datalad/distributed/drop.py     209      7    97%   189-190, 324-331, 403, 562, 716
datalad/local/remove.py          56      0   100%
-----------------------------------------------------------
TOTAL                           265      7    97%
----------------------------------------------------------------------
Ran 30 tests in 317.609s

FAILED (failures=3)
```

3x slower, some test assumptions broken.

mih commented:
Consolidated the tests and ported for cross-platform compatibility. Now approximately same speed and identical near-complete coverage:

Good FS
```
Name                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------
datalad/distributed/drop.py     209      5    98%   197, 324-331, 403, 607
datalad/local/remove.py          56      1    98%   142
-----------------------------------------------------------
TOTAL                           265      6    98%
----------------------------------------------------------------------
Ran 22 tests in 56.195s

OK
```

Crippled FS:

```
Name                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------
datalad/distributed/drop.py     209      5    98%   197, 324-331, 403, 607
datalad/local/remove.py          56      1    98%   142
-----------------------------------------------------------
TOTAL                           265      6    98%
----------------------------------------------------------------------
Ran 22 tests in 77.139s

OK
```

bpoldrack commented:
Not a huge issue, since it's not destroying data, but I wonder whether the availability check for revisions is ideal.
Example: I have a dataset w/ a subdataset. I clone that thing and in the clone I remove `origin` from the subdataset. If I now want to drop that subds from the clone, `drop` is complaining about no known remote. On the one hand, that's technically right, since that dataset doesn't know about a remote. On the other hand: If I drop it with `--reckless availability`, I can `get` it again w/o issues. At some level we know that it's fine. Looks to me as if the check is running on the knowledge of the to-be-dropped dataset only, while the reverse operation (`get`) would use the info as available from the superdataset.

Don't immediately see a better solution, but someone else might.

```
(datalad-dev) ben@tree in /tmp/test_drop_clone/subds on git:master
❱ git remote remove origin
(datalad-dev) ben@tree in /tmp/test_drop_clone/subds on git:master
❱ cd ..
(datalad-dev) ben@tree in /tmp/test_drop_clone on git:master
❱ datalad -f json_pp drop -d. subds --what datasets                                                                                                                                                            1 !
{
  "action": "uninstall",
  "message": [
    "to-be-dropped dataset has revisions that are not available at any known sibling. Use `datalad push --to ...` to push these before dropping the local dataset, or ignore via `--reckless availability`. Unique revisions: %s",
    [
      "master"
    ]
  ],
  "path": "/tmp/test_drop_clone/subds",
  "refds": "/tmp/test_drop_clone",
  "status": "error",
  "type": "dataset"
}
(datalad-dev) ben@tree in /tmp/test_drop_clone on git:master
❱ datalad -f json_pp drop -d. subds --what datasets --reckless availability                                                                                                                                    1 !
{
  "action": "uninstall",
  "path": "/tmp/test_drop_clone/subds",
  "refds": "/tmp/test_drop_clone",
  "status": "ok",
  "type": "dataset"
}
(datalad-dev) ben@tree in /tmp/test_drop_clone on git:master
❱ datalad get subds
[INFO   ] scanning for unlocked files (this may take some time)                                                                                                                                                    
install(ok): /tmp/test_drop_clone/subds (dataset) [Installed subdataset in order to get /tmp/test_drop_clone/subds]
get(ok): subds/another (file) [from origin...]                                                                                                                                                                     
action summary:                                                                                                                                                                                                    
  get (ok: 1)
  install (ok: 1)
```

bpoldrack commented:
Dropping a regular subdirectory w/ `--what datasets` returns absolutely nothing. No `notneeded`, no error - just silence.

```
❱ datalad -f json_pp drop somedir --what datasets
```

bpoldrack commented:
Similar: Dropping an already dropped subds, doesn't return anything either. Shoudn't that be a `notneeded`?

loj commented on datalad/distributed/drop.py:
```diff
@@ -0,0 +1,804 @@
+            Especially the latter is a potentially expensive check which might
+            involve numerous network transactions.
+            With 'undead', detection whether a to-be-removed local annex is
+            still known to exists in the network of dataset-clones is disabled.
```
A few small typos in this sentence:

"With 'undead', detection of whether a to-be-removed local annex is still known to exist in the network of dataset-clones is disabled."

add "of" and change "exists" -> "exist"

mih replied:
Fixed locally. Will push when nothing else is reported. Thx!

mih commented:
> Dropping a regular subdirectory w/ `--what datasets` returns absolutely nothing. No `notneeded`, no error - just silence.

Doing this for this simple case is easy, for the general case not. How would you determine that for any given input path whether it had absolutely no consequences -- keeping in mind:

- recursive operation
- multiple path arguments
- multiple path arguments pointing to different datasets

The desire for this type of reporting originally led to `AnnotatePaths`. We should keep in mind that `notneeded` results are inconsequential in general, and typically not even displayed. I doubt that it is worth complicating the code for it, or even invest runtime to determine whether this is happinging.

Same here ...

> Similar: Dropping an already dropped subds, doesn't return anything either. Shoudn't that be a notneeded?

How much time should we invest? An empty dir could be an uninstalled subdataset, or an empty dir. A non-existing path could be an uninstalled subdataset, or not exist at all. Do we report notneeded for first-level uninstalled datasets? But why not for next-level? -- they also do not exist. So generating these inconsequential results always require inspection, slow normal operations, and complicate the code, because we need to track numerous combinations when a report qualifies as "having said something already". The current implementation reports notneeded whenever it is cheap, or when required for backward compatibility. But it tries hard to avoid runtime costs for beautification.

> Example: I have a dataset w/ a subdataset. I clone that thing and in the clone I remove origin from the subdataset. If I now want to drop that subds from the clone, drop is complaining about no known remote. On the one hand, that's technically right, since that dataset doesn't know about a remote. On the other hand: If I drop it with --reckless availability, I can get it again w/o issues. At some level we know that it's fine. Looks to me as if the check is running on the knowledge of the to-be-dropped dataset only, while the reverse operation (get) would use the info as available from the superdataset.

What are you proposing? That on-drop the superdataset is inspected whether it can generate any candidate URL from which a repository can be cloned that is then inspected whether it has all the commits needed by the to-be-dropped subdataset?

loj commented:
This feels good! I tried it out for a couple of different workflows/situations that I commonly used drop/uninstall/remove, and I have no complaints. :-)

I just found the small typo in the docs for `--reckless undead`.

bpoldrack commented:
> How would you determine for any given input path whether it had absolutely no consequences

I might be wrong, but shouldn't such a case be discoverable right after determining `paths_by_ds`? A given path pointing to a regular subdir should be associated with `dataset`, rather than generate its own entry, right? So, `_drop_dataset` should be able to tell. In fact, I think right here: 

https://github.com/datalad/datalad/pull/6111/files#diff-f0ac845bc6391c474eb58b0c1448b7f6bc00cb5585dab58777d8d9512f99c06bR371

(link doesn't work properly. Line 371 in that diff's `drop.py`)

Comment right above that `return` seems to realize that basically. We have a path of which we know that ony possibly keys can be dropped, but `what` was `datasets`, so we don't do anything. Fine. Why not make that a `notneeded` result with respective message?

bpoldrack commented on datalad/distributed/drop.py:
```diff
@@ -0,0 +1,804 @@
+        # so we have paths constraints that prevent dropping the full dataset
+        # there is nothing to do here, but to drop keys, which we must not
+        # done
+        return
```
@mih

Something along these lines is what I mean. When do we end up here? AFAICS, when we have path(s) that are part of THIS dataset (as opposed to a subdataset) and we have `what='datasets'`. So, phrasing it like the comment ("constraining the drop of THIS dataset") is one way, but "trying to drop subdataset(s) where there are none" is another.

May be I just don't see how we can end up here in a way that this message would be misleading. I don't know.

```suggestion
        for p in paths:
            yield dict(action="drop",
                       path=p,
                       status="impossible",
                       message=f"what=datasets was given, but no dataset can be dropped from {p}",
                       type="file")
        return
```

mih replied:
I am afraid this is not correct. It is perfectly possible that the paths you are reporting on have been originally used to drop datasets underneath it. `impossible` would be wrong, because something might have been done already, and `file` is a guess.

bpoldrack replied:
> It is perfectly possible that the paths you are reporting on have been originally used to drop datasets underneath it.

Hm. I fail to produce that, but thanks for the hint. Trying to figure that out.

bpoldrack commented:
FTR: I'm testing my suggestion.

As for this:

> What are you proposing? That on-drop the superdataset is inspected whether it can generate any candidate URL from which a repository can be cloned that is then inspected whether it has all the commits needed by the to-be-dropped subdataset?

No - too expensive, obviously. As I said: No immediate idea. Just stumbled upon it. If you don't see a solution either - it's fine with me as is. I agree it's not worth making things more expensive.

> We should keep in mind that notneeded results are inconsequential in general, and typically not even displayed. I doubt that it is worth complicating the code for it, or even invest runtime to determine whether this is happinging.

Agree with the latter. Wouldn't phrase the former exactly like that. Via python interface I'd consider `notneeded` results much nicer to deal with than having to account for results sometimes, exceptions in other cases and absolutely nothing as yet another way of how datalad commands can behave. So - not worth paying a significant cost, but worth giving a second thought from my POV.

mih commented:
Conclusion from chat: Let's put the exploration and improvements possibly coming from @bpoldrack's comments in a separate PR. This one is already huge, and the suggested behavior was also not exhibited by the implementation that is being replaced it.
